gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAChallengeEndpoint.java

```diff
@@ -247,7 +258,29 @@ private void verifyCode(RoutingContext routingContext) {
 
         final FactorContext factorCtx = new FactorContext(applicationContext, factorData);
 
-        verify(factorProvider, factorCtx, h -> {
+        verifyAttemptService.checkVerifyAttempt(endUser.getId(), factorId, client, domain).subscribe(verifyAttempt -> {
+                    if (verifyAttempt.isAllowRequest()) {
+                        verify(factorProvider, factorCtx, Optional.of(verifyAttempt), routingContext,
+                                verifyHandler(routingContext, client, endUser, factor, code, factorId, factorProvider, enrolledFactor));
+                    } else {
+                        if (verifyAttemptService.shouldSendEmail(client, domain)) {
+                            new Thread(() -> emailService.send(Template.VERIFY_ATTEMPT, endUser, client)).start();
+                        }
+                        handleException(routingContext, VERIFY_ATTEMPT_ERROR_PARAM_KEY, "maximum_verify_limit");
+                    }
+                },
+                error -> {
+                    logger.error("Could not check verify attempts", error);
+                    routingContext.fail(401);
+                },
+                () -> verify(factorProvider, factorCtx, Optional.empty(), routingContext,
+                        verifyHandler(routingContext, client, endUser, factor, code, factorId, factorProvider, enrolledFactor)));
```

Comment:
This maybe useless if we revisit the `checkVerifyAttempt` to return a `Single<VerifyAttempt>` instead of `Maybe<VerifyAttempt>`. Let me think about it :thinking:

Reply:
This one I am not sure, we can discuss

Reply:
ok I think I got it.

IMHO, the `checkVerifyAttempt` should perform the check on the `isAllowRequest`, if false, then the Maybe return by this method will contain an Error (New Exeption like MFAVerificationAttemptException).

based on that:
* the `onSuccess` method of the subscription will on call the `verify` method.
* the `onError` method will test the type of excepion, if it is a MFAValidationAttempt, we can reurn the right error code and send the email if expected, otherwise, we do the same as previously


```
        verifyAttemptService.checkVerifyAttempt(endUser.getId(), factorId, client, domain)
                .switchIfEmpty(Optional::empty)
                .subscribe(verifyAttempt -> {
                            final var verifyResultHandler = verifyHandler(routingContext, client, endUser, factor, code, factorId, factorProvider, enrolledFactor);
                            verify(factorProvider, factorCtx, verifyAttempt, routingContext, verifyResultHandler);
                },
                error -> {
                    if (error instanceof MFAValidatioAttemptException) {
                        if (verifyAttemptService.shouldSendEmail(client, domain)) {
                            emailService.send(Template.VERIFY_ATTEMPT, endUser, client);
                        }
                        handleException(routingContext, VERIFY_ATTEMPT_ERROR_PARAM_KEY, "maximum_verify_limit");
                    } else {
                        logger.error("Could not check verify attempts", error);
                        routingContext.fail(401);
                    }
                });
```

Reply:
done